File: frontend/src/types/gm.ts
// GM Roles - matches database enum
export type GMRole = 'helper' | 'moderator' | 'game_master' | 'admin' | 'owner';

// Ban Types
export type BanType = 'permanent' | 'temporary' | 'ip' | 'hwid';

// Keep GMLevel for backwards compatibility with UI components
export type GMLevel = 1 | 2 | 3 | 4 | 5;

export const GM_ROLE_NAMES: Record<GMRole, string> = {
  helper: 'Helper',
  moderator: 'Moderator',
  game_master: 'Game Master',
  admin: 'Admin',
  owner: 'Owner',
};

export const GM_LEVEL_NAMES: Record<GMLevel, string> = {
  1: 'Helper',
  2: 'Moderator',
  3: 'Game Master',
  4: 'Admin',
  5: 'Owner',
};

export const gmRoleToLevel = (role: GMRole): GMLevel => {
  const map: Record<GMRole, GMLevel> = {
    helper: 1,
    moderator: 2,
    game_master: 3,
    admin: 4,
    owner: 5,
  };
  return map[role];
};

// GM Permissions
export interface GMPermissions {
  can_ban: boolean;
  can_mute: boolean;
  can_jail: boolean;
  can_kick: boolean;
  can_teleport: boolean;
  can_spawn_items: boolean;
  can_modify_gold: boolean;
  can_view_tickets: boolean;
  can_send_announcements: boolean;
  can_view_player_data: boolean;
  can_modify_player_data: boolean;
  can_access_market_free: boolean;
  can_invisible: boolean;
  can_god_mode: boolean;
}

export interface GMAccount {
  id: string;
  account_id: string;
  gm_role: GMRole;
  gm_name: string;
  permissions?: GMPermissions;
  is_active: boolean;
  is_visible: boolean;
  is_on_duty: boolean;
  created_at: string;
  last_action_at?: string;
  // Computed for UI convenience
  level?: GMLevel;
}

export interface GMAuthResponse {
  token: string;
  gm_account: GMAccount;
}

// Ban - account level
export interface Ban {
  id: string;
  account_id: string;
  ban_type: string;
  reason: string;
  banned_by: string;
  starts_at: string;
  expires_at?: string;
  is_active: boolean;
  unbanned_by?: string;
  unbanned_at?: string;
  unban_reason?: string;
  banned_ip?: string;
  created_at: string;
}

export interface BanRequest {
  account_id: string;
  ban_type: string;
  reason: string;
  duration?: number; // hours
}

// Mute - character level
export interface Mute {
  id: string;
  character_id: string;
  mute_type: string;
  reason: string;
  muted_by: string;
  starts_at: string;
  expires_at: string;
  is_active: boolean;
  created_at: string;
}

export interface MuteRequest {
  character_id: string;
  mute_type: string;
  reason: string;
  duration: number; // minutes
}

// Announcement
export type AnnouncementType = 'global' | 'server' | 'maintenance' | 'event';

export interface Announcement {
  id: string;
  server_id?: number;
  announcement_type: AnnouncementType;
  title?: string;
  message: string;
  show_in_chat: boolean;
  show_as_popup: boolean;
  show_in_ticker: boolean;
  color: string;
  icon?: string;
  created_by?: string;
  starts_at: string;
  expires_at?: string;
  is_active: boolean;
  created_at: string;
  // Compatibility aliases
  type?: AnnouncementType;
  content?: string;
}

export interface AnnouncementRequest {
  server_id?: number;
  announcement_type: AnnouncementType;
  title?: string;
  message: string;
  show_in_chat: boolean;
  show_as_popup: boolean;
  color: string;
  expires_at?: string;
  // Compatibility aliases
  type?: AnnouncementType;
  content?: string;
}

// Ticket - character based
export type TicketStatus = 'open' | 'in_progress' | 'resolved' | 'closed';
export type TicketCategory = 'bug' | 'report' | 'support' | 'feedback' | 'appeal';
export type TicketPriority = 'low' | 'medium' | 'high' | 'critical';

export interface Ticket {
  id: string;
  server_id: number;
  reporter_id: string;
  category: string;
  priority: string;
  subject: string;
  description: string;
  related_character_id?: string;
  status: TicketStatus;
  assigned_to?: string;
  assigned_at?: string;
  resolution?: string;
  resolved_by?: string;
  resolved_at?: string;
  created_at: string;
  updated_at: string;
}

export interface CreateTicketRequest {
  server_id: number;
  category: string;
  priority: string;
  subject: string;
  description: string;
  related_character_id?: string;
}

// Ticket Message
export interface TicketMessage {
  id: string;
  ticket_id: string;
  sender_type: 'player' | 'gm';
  sender_character_id?: string;
  sender_gm_id?: string;
  message: string;
  is_internal: boolean;
  created_at: string;
}

export interface TicketMessageRequest {
  message: string;
  is_internal?: boolean;
}

// Legacy alias for compatibility
export interface TicketResponse extends TicketMessage {}

// Dashboard Stats
export interface DashboardStats {
  total_accounts: number;
  online_characters: number;
  active_bans: number;
  open_tickets: number;
  total_characters: number;
  active_mutes: number;
}

// Private Message - between characters
export interface PrivateMessage {
  id: string;
  sender_id: string;
  recipient_id: string;
  message: string;
  is_read: boolean;
  read_at?: string;
  is_gm_message: boolean;
  deleted_by_sender: boolean;
  deleted_by_recipient: boolean;
  created_at: string;
  // Computed fields with names
  sender_name?: string;
  recipient_name?: string;
  // Legacy compatibility
  receiver_id?: string;
  receiver_name?: string;
  content?: string;
}

export interface SendMessageRequest {
  recipient_id: string;
  message: string;
}

File: backend/internal/models/gm.go
package models

import (
	"time"

	"github.com/google/uuid"
)

// GM Roles - matches database enum
type GMRole string

const (
	GMRoleHelper     GMRole = "helper"
	GMRoleModerator  GMRole = "moderator"
	GMRoleGameMaster GMRole = "game_master"
	GMRoleAdmin      GMRole = "admin"
	GMRoleOwner      GMRole = "owner"
)

// GMRole to level for permission checks
func (r GMRole) Level() int {
	switch r {
	case GMRoleHelper:
		return 1
	case GMRoleModerator:
		return 2
	case GMRoleGameMaster:
		return 3
	case GMRoleAdmin:
		return 4
	case GMRoleOwner:
		return 5
	default:
		return 0
	}
}

// GM Permissions - stored as JSONB
type GMPermissions struct {
	CanBan               bool `json:"can_ban"`
	CanMute              bool `json:"can_mute"`
	CanJail              bool `json:"can_jail"`
	CanKick              bool `json:"can_kick"`
	CanTeleport          bool `json:"can_teleport"`
	CanSpawnItems        bool `json:"can_spawn_items"`
	CanModifyGold        bool `json:"can_modify_gold"`
	CanViewTickets       bool `json:"can_view_tickets"`
	CanSendAnnouncements bool `json:"can_send_announcements"`
	CanViewPlayerData    bool `json:"can_view_player_data"`
	CanModifyPlayerData  bool `json:"can_modify_player_data"`
	CanAccessMarketFree  bool `json:"can_access_market_free"`
	CanInvisible         bool `json:"can_invisible"`
	CanGodMode           bool `json:"can_god_mode"`
}

// GM Account - linked to accounts table via account_id
type GMAccount struct {
	ID           uuid.UUID      `json:"id"`
	AccountID    uuid.UUID      `json:"account_id"`
	GMRole       GMRole         `json:"gm_role"`
	GMName       string         `json:"gm_name"`
	Permissions  *GMPermissions `json:"permissions,omitempty"`
	IsActive     bool           `json:"is_active"`
	IsVisible    bool           `json:"is_visible"`
	IsOnDuty     bool           `json:"is_on_duty"`
	CreatedAt    time.Time      `json:"created_at"`
	LastActionAt *time.Time     `json:"last_action_at,omitempty"`
}

// GM login uses normal account credentials
type GMLoginRequest struct {
	Email    string `json:"email"`
	Password string `json:"password"`
}

type GMAuthResponse struct {
	Token     string     `json:"token"`
	GMAccount *GMAccount `json:"gm_account"`
}

// Ban System
type Ban struct {
	ID          uuid.UUID  `json:"id"`
	AccountID   uuid.UUID  `json:"account_id"`
	BanType     string     `json:"ban_type"`
	Reason      string     `json:"reason"`
	BannedBy    uuid.UUID  `json:"banned_by"`
	StartsAt    time.Time  `json:"starts_at"`
	ExpiresAt   *time.Time `json:"expires_at,omitempty"`
	IsActive    bool       `json:"is_active"`
	UnbannedBy  *uuid.UUID `json:"unbanned_by,omitempty"`
	UnbannedAt  *time.Time `json:"unbanned_at,omitempty"`
	UnbanReason *string    `json:"unban_reason,omitempty"`
	BannedIP    *string    `json:"banned_ip,omitempty"`
	CreatedAt   time.Time  `json:"created_at"`
}

type BanRequest struct {
	AccountID uuid.UUID `json:"account_id"`
	BanType   string    `json:"ban_type"`
	Reason    string    `json:"reason"`
	Duration  *int      `json:"duration,omitempty"`
}

// Mute System - character_id based
type Mute struct {
	ID          uuid.UUID `json:"id"`
	CharacterID uuid.UUID `json:"character_id"`
	MuteType    string    `json:"mute_type"`
	Reason      string    `json:"reason"`
	MutedBy     uuid.UUID `json:"muted_by"`
	StartsAt    time.Time `json:"starts_at"`
	ExpiresAt   time.Time `json:"expires_at"`
	IsActive    bool      `json:"is_active"`
	CreatedAt   time.Time `json:"created_at"`
}

type MuteRequest struct {
	CharacterID uuid.UUID `json:"character_id"`
	MuteType    string    `json:"mute_type"`
	Reason      string    `json:"reason"`
	Duration    int       `json:"duration"`
}

// Announcement System
type Announcement struct {
	ID               uuid.UUID  `json:"id"`
	ServerID         *int       `json:"server_id,omitempty"`
	AnnouncementType string     `json:"announcement_type"`
	Title            *string    `json:"title,omitempty"`
	Message          string     `json:"message"`
	ShowInChat       bool       `json:"show_in_chat"`
	ShowAsPopup      bool       `json:"show_as_popup"`
	ShowInTicker     bool       `json:"show_in_ticker"`
	Color            string     `json:"color"`
	Icon             *string    `json:"icon,omitempty"`
	CreatedBy        *uuid.UUID `json:"created_by,omitempty"`
	StartsAt         time.Time  `json:"starts_at"`
	ExpiresAt        *time.Time `json:"expires_at,omitempty"`
	IsActive         bool       `json:"is_active"`
	CreatedAt        time.Time  `json:"created_at"`
}

type AnnouncementRequest struct {
	ServerID         *int       `json:"server_id,omitempty"`
	AnnouncementType string     `json:"announcement_type"`
	Title            *string    `json:"title,omitempty"`
	Message          string     `json:"message"`
	ShowInChat       bool       `json:"show_in_chat"`
	ShowAsPopup      bool       `json:"show_as_popup"`
	Color            string     `json:"color"`
	ExpiresAt        *time.Time `json:"expires_at,omitempty"`
}

// Ticket System - character based (reporter_id)
type Ticket struct {
	ID                 uuid.UUID  `json:"id"`
	ServerID           int        `json:"server_id"`
	ReporterID         uuid.UUID  `json:"reporter_id"`
	Category           string     `json:"category"`
	Priority           string     `json:"priority"`
	Subject            string     `json:"subject"`
	Description        string     `json:"description"`
	RelatedCharacterID *uuid.UUID `json:"related_character_id,omitempty"`
	Status             string     `json:"status"`
	AssignedTo         *uuid.UUID `json:"assigned_to,omitempty"`
	AssignedAt         *time.Time `json:"assigned_at,omitempty"`
	Resolution         *string    `json:"resolution,omitempty"`
	ResolvedBy         *uuid.UUID `json:"resolved_by,omitempty"`
	ResolvedAt         *time.Time `json:"resolved_at,omitempty"`
	CreatedAt          time.Time  `json:"created_at"`
	UpdatedAt          time.Time  `json:"updated_at"`
}

type CreateTicketRequest struct {
	ServerID           int        `json:"server_id"`
	Category           string     `json:"category"`
	Priority           string     `json:"priority"`
	Subject            string     `json:"subject"`
	Description        string     `json:"description"`
	RelatedCharacterID *uuid.UUID `json:"related_character_id,omitempty"`
}

// Ticket Message
type TicketMessage struct {
	ID                uuid.UUID  `json:"id"`
	TicketID          uuid.UUID  `json:"ticket_id"`
	SenderType        string     `json:"sender_type"`
	SenderCharacterID *uuid.UUID `json:"sender_character_id,omitempty"`
	SenderGMID        *uuid.UUID `json:"sender_gm_id,omitempty"`
	Message           string     `json:"message"`
	IsInternal        bool       `json:"is_internal"`
	CreatedAt         time.Time  `json:"created_at"`
}

type TicketMessageRequest struct {
	Message    string `json:"message"`
	IsInternal bool   `json:"is_internal,omitempty"`
}

// Private Message - between characters
type PrivateMessage struct {
	ID                 uuid.UUID  `json:"id"`
	SenderID           uuid.UUID  `json:"sender_id"`
	RecipientID        uuid.UUID  `json:"recipient_id"`
	Message            string     `json:"message"`
	IsRead             bool       `json:"is_read"`
	ReadAt             *time.Time `json:"read_at,omitempty"`
	IsGMMessage        bool       `json:"is_gm_message"`
	DeletedBySender    bool       `json:"deleted_by_sender"`
	DeletedByRecipient bool       `json:"deleted_by_recipient"`
	CreatedAt          time.Time  `json:"created_at"`
}

type SendMessageRequest struct {
	RecipientID uuid.UUID `json:"recipient_id"`
	Message     string    `json:"message"`
}

type MessageWithNames struct {
	PrivateMessage
	SenderName    string `json:"sender_name"`
	RecipientName string `json:"recipient_name"`
}

// Dashboard Stats
type DashboardStats struct {
	TotalAccounts    int `json:"total_accounts"`
	OnlineCharacters int `json:"online_characters"`
	ActiveBans       int `json:"active_bans"`
	OpenTickets      int `json:"open_tickets"`
	TotalCharacters  int `json:"total_characters"`
	ActiveMutes      int `json:"active_mutes"`
}
